ledger/src/shred/wire.rs

```diff
@@ -21,36 +22,39 @@ use {
 };
 
 #[inline]
-fn get_shred_size(packet: &Packet) -> Option<usize> {
-    let size = packet.data(..)?.len();
-    if packet.meta().repair() {
-        size.checked_sub(SIZE_OF_NONCE)
-    } else {
-        Some(size)
-    }
+fn get_shred_size(shred: &[u8]) -> Option<usize> {
+    // Legacy data shreds have zero padding at the end which might have been
+    // trimmed. Other variants do not have any trailing zeros.
+    Some(match get_shred_variant(shred).ok()? {
+        ShredVariant::LegacyCode => shred::legacy::ShredCode::SIZE_OF_PAYLOAD,
+        ShredVariant::LegacyData => shred::legacy::ShredData::SIZE_OF_PAYLOAD.min(shred.len()),
+        ShredVariant::MerkleCode { .. } => shred::merkle::ShredCode::SIZE_OF_PAYLOAD,
+        ShredVariant::MerkleData { .. } => shred::merkle::ShredData::SIZE_OF_PAYLOAD,
+    })
 }
 
 #[inline]
 pub fn get_shred(packet: &Packet) -> Option<&[u8]> {
-    let size = get_shred_size(packet)?;
-    packet.data(..size)
+    let data = packet.data(..)?;
+    data.get(..get_shred_size(data)?)
 }
 
 #[inline]
 pub fn get_shred_mut(packet: &mut Packet) -> Option<&mut [u8]> {
-    let size = get_shred_size(packet)?;
-    packet.buffer_mut().get_mut(..size)
+    let buffer = packet.buffer_mut();
+    buffer.get_mut(..get_shred_size(buffer)?)
 }
 
 #[inline]
 pub fn get_shred_and_repair_nonce(packet: &Packet) -> Option<(&[u8], Option<Nonce>)> {
     let data = packet.data(..)?;
+    let shred = data.get(..get_shred_size(data)?)?;
```

Comment:
nit: Looks like you could use `get_shred()` here instead of repeating this line and the line above

Reply:
Need `data` below, to get nonce out of it.
`get_shred` excludes the `nonce` part.
So I still need to call `packet.data(..)?`, and if I use `get_shred` there will be 2 calls to `packet.data(..)?` which is kind of wasteful.

Reply:
D'oh, you're right